### PR TITLE
scripts: runners: improve hex file detection and output

### LINKED ISSUE #18201: bug: west flash with --hex-file param used to work w/o path specified
**Describe the bug**
At some point during Zephyr v2.0 development, `west sign` moved the zephyr.signed.hex(.bin) into the build dir.  However, users jumping from v1.14 to v2.0 who use their old commands like:
`west flash -d build/lwm2m-nrf52840_pca10056 --hex-file zephyr.signed.hex` will be faced with very cryptic failures when `zephyr.signed.hex` is silently skipped because it's no longer present at the root of the Zephyr west dir and the `zephyr.bin` file in the build dir is used instead.

**To Reproduce**
Steps to reproduce the behavior (most likely to happen when building a firmware for use with mcuboot):
1. west build -d build/lwm2m-nrf52840_pca10056 -b nrf52840_pca10056 -s zephyr/samples/net/lwm2m_client -- -DOVERLAY_CONFIG=overlay-ieee802154.conf -DCONFIG_BOOTLOADER_MCUBOOT=y
2. west sign  -d build/lwm2m-nrf52840_pca10056 -t imgtool -- --key mcuboot/root-rsa-2048.pem
3. west flash -d build/lwm2m-nrf52840_pca10056 --hex-file zephyr.signed.hex

Response from UART is:
```
[00:00:00.041,168] <err> mcuboot: Image in the primary slot is not valid!                                                             
[00:00:00.048,706] <err> mcuboot: Unable to find bootable image                                                                       
```

**Expected behavior**
Users migrating from v1.14 to v2.0 will expect this to work or complain with an error, but not silently flash the wrong binary.

**Impact**
I expect this will be very confusing.

**Environment (please complete the following information):**
 - OS: Arch Linux
 - Toolchain: Zephyr SDK
 - Commit SHA or Version used: Zephyr v2.0.0-rc1

### PR DESCRIPTION
**Pull request by mbolivar**
This is a band-aid to make it more obvious to potential users of 'west
sign' and 'west flash' which hex file they are flashing, when they are
falling back on a binary file, and erroring out when a hex file does
not exist.

Fixes: #18201

### CODE REVIEW

**Comment by mbolivar** [comment]
This is a generalization and slight tweak of c484935fe5dc2fc93f2b4c54ba85807aba15b171.

**Review by galak — APPROVED** [review]
LGTM.  I wondering if you could move the logic for which file formats are required up a level and than just have the specific runner specify which types.

**Review by carlescufi — APPROVED** [review]
LGTM as a hotfix

**Comment by mbolivar** [comment]
> LGTM. I wondering if you could move the logic for which file formats are required up a level and than just have the specific runner specify which types.

The real fix IMO is to deprecate --bin-file, --hex-file, and --elf-file entirely and just use a single --file argument. I expect centralizing what type of file you can flash within RunnerCaps around that time will make sense.

**Comment by mike-scott** [comment]
@mbolivar so the goal with this PR isn't to make the commands work like they did pre v2.0.  It's more to explain what's being flashed so that the user has a better understanding?

**Comment by mike-scott** [comment]
Making sure I can build the 4.0.4 userspace in Alpine, before I merge this.

**Comment by mbolivar** [comment]
> @mbolivar so the goal with this PR isn't to make the commands work like they did pre v2.0.

Are you talking about `west sign`? From the original bug report,

> At some point during Zephyr v2.0 development, west sign moved the zephyr.signed.hex(.bin) into the build dir.

If so, the move of the default signed file location from `$PWD` to the build directory was on purpose and I would be strongly against changing it. It plays a lot better with west config options like `build.dir-fmt` that let you create multiple build directories depending on the board, app, etc.

You can always give the path to the output signed file explicitly, using the LTS or 2.0 versions of `west sign`, if you need consistent results.

I put the signed files in the current directory because I thought the build system maintainers wouldn't want me to be messing with files in there outside of CMake. Turns out I was wrong, so I made that change (also other users requested this change).

> It's more to explain what's being flashed so that the user has a better understanding?

Not just that. It also errors out if you pass an explicit `--hex-file` that does not exist, or the system is otherwise looking for a nonexistent hex file. In the original report, these were the steps to reproduce and problem:

```

    west build -d build/lwm2m-nrf52840_pca10056 -b nrf52840_pca10056 -s zephyr/samples/net/lwm2m_client -- -DOVERLAY_CONFIG=overlay-ieee802154.conf -DCONFIG_BOOTLOADER_MCUBOOT=y
    west sign -d build/lwm2m-nrf52840_pca10056 -t imgtool -- --key mcuboot/root-rsa-2048.pem
    west flash -d build/lwm2m-nrf52840_pca10056 --hex-file zephyr.signed.hex
```

The expected behavior was "Users migrating from v1.14 to v2.0 will expect this to work or complain with an error, but not silently flash the wrong binary."

The error should not be silently ignored now with these patches, except perhaps on a few obscure runners I don't want to touch :).

**Comment by ioannisg** [comment]
@mike-scott could you take another look at this converstation?

**Review by mike-scott — APPROVED** [review]
This looks good as a way of adding more information to the flash step.   And I approve of that.  Unfortunately, all of the developers using MCUboot moving from v1.14 to v2.0 will still need to change their flashing work flow which is undesirable.  This means documentation, CI and other things are all going to need retooling.

But, for now, that's "just what happens with Zephyr".

**Pull request merged**
